libs/sdk-core/src/lsps0/transport.rs

```diff
@@ -0,0 +1,677 @@
+use std::sync::Arc;
+use std::time::Duration;
+
+use crate::CustomMessage;
+use crate::NodeAPI;
+use anyhow::{anyhow, Result};
+use rand::distributions::Alphanumeric;
+use rand::distributions::DistString;
+use serde::de::DeserializeOwned;
+use tokio::sync::{mpsc, oneshot, Mutex, RwLock};
+
+use super::error::Error;
+use super::jsonrpc::{
+    JsonRpcError, JsonRpcNotification, JsonRpcRequest, JsonRpcResponse, JsonRpcServerObject,
+};
+
+const LSPS0_MESSAGE_TYPE: u16 = 37913;
+const JSONRPC_VERSION: &str = "2.0";
+
+#[tonic::async_trait]
+trait NotificationSender: Send + Sync {
+    async fn send(&self, msg: JsonRpcNotification) -> Result<(), Error>;
+}
+struct NotificationHandler<TNotification>
+where
+    TNotification: DeserializeOwned + Send,
+{
+    tx: mpsc::Sender<TNotification>,
+}
+
+#[tonic::async_trait]
+impl<TNotification> NotificationSender for NotificationHandler<TNotification>
+where
+    TNotification: DeserializeOwned + Send,
+{
+    async fn send(&self, msg: JsonRpcNotification) -> Result<(), Error> {
+        let n = match serde_json::from_value::<TNotification>(msg.params) {
+            Ok(n) => n,
+            Err(e) => return Err(Error::Deserialization(e)),
+        };
+
+        match self.tx.send(n).await {
+            Ok(_) => Ok(()),
+            Err(_) => Err(Error::Local(anyhow!("receiver dropped"))),
+        }
+    }
+}
+
+struct ResponseOrError {
+    response: Option<JsonRpcResponse>,
+    error: Option<JsonRpcError>,
+}
+
+/// Transport sends and receives LSPS0 messages to/from remote nodes. One
+/// user node has one Transport.
+pub struct Transport {
+    node: Arc<dyn NodeAPI>,
+    response_handlers: Mutex<HashMap<String, oneshot::Sender<ResponseOrError>>>,
+    notification_handlers: RwLock<HashMap<String, Box<dyn NotificationSender>>>,
+}
+
+impl Transport {
+    #[allow(dead_code)]
+    pub fn new(node: Arc<dyn NodeAPI>) -> Transport {
+        Transport {
+            node,
+            response_handlers: Mutex::new(HashMap::new()),
+            notification_handlers: RwLock::new(HashMap::new()),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn start(&self) -> Result<()> {
+        debug!("starting lsps0 tranport.");
+        let mut stream = self.node.stream_custom_messages().await?;
+        loop {
+            let msg = match stream.recv().await {
+                Some(val) => val,
+                None => break Err(anyhow!("custom message stream ended.")),
+            };
+            if msg.message_type != LSPS0_MESSAGE_TYPE {
+                continue;
+            }
+
+            let v: JsonRpcServerObject = match serde_json::from_slice(&msg.payload) {
+                Ok(v) => v,
+                Err(e) => {
+                    debug!(
+                        "error deserializing lsps0 payload {:?}: {}",
+                        &msg.payload, e
+                    );
+                    continue;
+                }
+            };
+
+            let jsonrpc_version = match &v {
```

Comment:
This match scope hints that perhaps we need an outer structure with the common fields?
For example:

```
struct JsonPayload {
  jsonrpc: string,

  #[serde(flatten)]
  object: JSONRpcServerObject,
}
```

The flatten means the fields are embedded.
Only if it makes sense, didn't think about it all the way through.

Reply:
Makes sense, done